File: packages/hht_auth/hht_auth_server/lib/src/services/jwt_service.dart
/// IMPLEMENTS REQUIREMENTS:
///   REQ-d00078: HHT Diary Auth Service - JWT generation and verification
///
/// JWT service for generating and verifying RS256-signed JSON Web Tokens.
///
/// Tokens are short-lived (15 minutes) and contain user identification,
/// sponsor information, and session data. The service uses RSA key pairs
/// for secure signing and verification.

import 'package:dart_jsonwebtoken/dart_jsonwebtoken.dart';
import 'package:hht_auth_core/hht_auth_core.dart';

/// Service for JWT token operations using RS256 algorithm.
class JwtService {
  final String _privateKey;
  final String _publicKey;
  final String _issuer;

  /// Creates a JWT service with RSA key pair.
  ///
  /// [privateKey] PEM-encoded RSA private key for signing tokens
  /// [publicKey] PEM-encoded RSA public key for verifying tokens
  /// [issuer] Token issuer identifier (e.g., 'hht-auth-service')
  JwtService({
    required String privateKey,
    required String publicKey,
    required String issuer,
  })  : _privateKey = privateKey,
        _publicKey = publicKey,
        _issuer = issuer;

  /// Generates a JWT string from an AuthToken model.
  ///
  /// Returns a signed JWT string with RS256 algorithm.
  String generateToken(AuthToken authToken) {
    final jwt = JWT(
      {
        'sub': authToken.sub,
        'username': authToken.username,
        'sponsorId': authToken.sponsorId,
        'sponsorUrl': authToken.sponsorUrl,
        'appUuid': authToken.appUuid,
        'iat': authToken.iat.millisecondsSinceEpoch ~/ 1000,
        'exp': authToken.exp.millisecondsSinceEpoch ~/ 1000,
      },
      issuer: _issuer,
    );

    return jwt.sign(
      RSAPrivateKey(_privateKey),
      algorithm: JWTAlgorithm.RS256,
    );
  }

  /// Verifies a JWT string and returns the decoded AuthToken.
  ///
  /// Returns null if the token is invalid, expired, or malformed.
  AuthToken? verifyToken(String token) {
    try {
      final jwt = JWT.verify(
        token,
        RSAPublicKey(_publicKey),
        issuer: _issuer,
      );

      final payload = jwt.payload as Map<String, dynamic>;

      // Validate required claims
      if (!payload.containsKey('sub') ||
          !payload.containsKey('username') ||
          !payload.containsKey('sponsorId') ||
          !payload.containsKey('sponsorUrl') ||
          !payload.containsKey('appUuid') ||
          !payload.containsKey('iat') ||
          !payload.containsKey('exp')) {
        return null;
      }

      return AuthToken(
        sub: payload['sub'] as String,
        username: payload['username'] as String,
        sponsorId: payload['sponsorId'] as String,
        sponsorUrl: payload['sponsorUrl'] as String,
        appUuid: payload['appUuid'] as String,
        iat: DateTime.fromMillisecondsSinceEpoch(
          (payload['iat'] as int) * 1000,
        ),
        exp: DateTime.fromMillisecondsSinceEpoch(
          (payload['exp'] as int) * 1000,
        ),
      );
    } on JWTExpiredException {
      return null;
    } on JWTException {
      return null;
    } catch (e) {
      return null;
    }
  }

  /// Refreshes an existing token with a new expiry time.
  ///
  /// Returns a new JWT string with the same claims but updated iat and exp.
  /// Returns null if the original token is invalid or expired.
  String? refreshToken(String token) {
    final decoded = verifyToken(token);
    if (decoded == null) return null;

    final now = DateTime.now();
    final newToken = decoded.copyWith(
      iat: now,
      exp: now.add(Duration(minutes: 15)),
    );

    return generateToken(newToken);
  }

  /// Extracts JWT token from an Authorization header.
  ///
  /// Expected format: "Bearer <token>"
  /// Returns the token string or null if the header is malformed.
  String? extractTokenFromHeader(String? header) {
    if (header == null || header.isEmpty) return null;

    final parts = header.split(' ');
    if (parts.length != 2 || parts[0] != 'Bearer') return null;

    final token = parts[1].trim();
    return token.isEmpty ? null : token;
  }
}

File: packages/hht_auth/hht_auth_server_ts/src/services/jwt.service.ts
/**
 * IMPLEMENTS REQUIREMENTS:
 *   REQ-d00XXX: JWT service for Web Diary authentication
 *
 * JwtService - JWT token generation and verification for authentication
 *
 * CRITICAL: This implementation MUST maintain 100% compatibility with the Dart client.
 * Any changes to token structure or claims MUST be coordinated with client updates.
 *
 * Key design decisions:
 * - Algorithm: RS256 (RSA with SHA-256) for asymmetric signing
 * - Timestamps: Unix seconds (NOT milliseconds) for iat/exp
 * - Default expiry: 15 minutes (900 seconds)
 * - Library: jose (recommended by Node.js community for JWT operations)
 */

import * as jose from 'jose';

/**
 * Configuration for creating JwtService from PEM strings
 */
export interface JwtConfig {
  privateKey: string;
  publicKey: string;
  issuer: string;
}

/**
 * JWT Payload structure - MUST match Dart client exactly
 */
export interface JwtPayload {
  sub: string; // User ID (UUID v4)
  username: string;
  sponsorId: string;
  sponsorUrl: string; // Sponsor Portal base URL
  appUuid: string; // App instance UUID
  iat?: number; // Unix timestamp in SECONDS
  exp?: number; // Unix timestamp in SECONDS
  iss?: string; // Issuer
}

/**
 * Input payload for token generation (without iat/exp/iss)
 */
export interface TokenPayloadInput {
  sub: string;
  username: string;
  sponsorId: string;
  sponsorUrl: string;
  appUuid: string;
}

/**
 * JwtService - Handles JWT token lifecycle
 *
 * Responsibilities:
 * - Generate signed JWT tokens with RS256 algorithm
 * - Verify token signature and expiry
 * - Refresh tokens with updated timestamps
 * - Extract tokens from Authorization headers
 */
export class JwtService {
  private readonly DEFAULT_EXPIRY_SECONDS = 900; // 15 minutes
  private readonly ALGORITHM = 'RS256' as const; // RSA with SHA-256
  private readonly BEARER_PREFIX = 'Bearer ';

  constructor(
    private readonly privateKey: jose.KeyLike,
    private readonly publicKey: jose.KeyLike,
    private readonly issuer: string
  ) {}

  /**
   * Create JwtService from PEM-encoded key strings
   *
   * @param config - Configuration with PEM strings and issuer
   * @returns JwtService instance
   */
  static async fromConfig(config: JwtConfig): Promise<JwtService> {
    const privateKey = await jose.importPKCS8(config.privateKey, 'RS256');
    const publicKey = await jose.importSPKI(config.publicKey, 'RS256');
    return new JwtService(privateKey, publicKey, config.issuer);
  }

  /**
   * Generate a signed JWT token
   *
   * @param payload - User claims (sub, username, sponsorId, sponsorUrl, appUuid)
   * @returns Signed JWT token string
   *
   * CRITICAL: Uses Unix SECONDS for iat/exp (not milliseconds)
   */
  async generateToken(payload: TokenPayloadInput): Promise<string> {
    const now = Math.floor(Date.now() / 1000); // Unix seconds, NOT milliseconds
    const exp = now + this.DEFAULT_EXPIRY_SECONDS;

    return await new jose.SignJWT({
      sub: payload.sub,
      username: payload.username,
      sponsorId: payload.sponsorId,
      sponsorUrl: payload.sponsorUrl,
      appUuid: payload.appUuid,
    })
      .setProtectedHeader({ alg: this.ALGORITHM })
      .setIssuedAt(now)
      .setExpirationTime(exp)
      .setIssuer(this.issuer)
      .sign(this.privateKey);
  }

  /**
   * Verify and decode a JWT token
   *
   * @param token - JWT token string to verify
   * @returns Decoded payload if valid, null if invalid/expired/tampered
   *
   * Returns null for ANY verification failure to prevent security issues
   */
  async verifyToken(token: string): Promise<JwtPayload | null> {
    if (!token || token.trim() === '') {
      return null;
    }

    try {
      const { payload } = await jose.jwtVerify(token, this.publicKey, {
        issuer: this.issuer,
      });

      // jose.JWTPayload is a superset that includes standard claims
      // We safely extract our custom claims
      return payload as unknown as JwtPayload;
    } catch (error) {
      // Any verification failure returns null:
      // - Expired token (jose.errors.JWTExpired)
      // - Invalid signature (jose.errors.JWSSignatureVerificationFailed)
      // - Wrong issuer (jose.errors.JWTClaimValidationFailed)
      // - Malformed token
      return null;
    }
  }

  /**
   * Refresh a token with updated timestamps
   *
   * @param token - Existing valid JWT token
   * @returns New token with same claims but fresh iat/exp, or null if input invalid
   *
   * Use case: Extend session without requiring re-authentication
   */
  async refreshToken(token: string): Promise<string | null> {
    // Verify the existing token first
    const payload = await this.verifyToken(token);

    if (!payload) {
      return null;
    }

    // Generate new token with same claims (except iat/exp)
    const newToken = await this.generateToken({
      sub: payload.sub,
      username: payload.username,
      sponsorId: payload.sponsorId,
      sponsorUrl: payload.sponsorUrl,
      appUuid: payload.appUuid,
    });

    return newToken;
  }

  /**
   * Extract JWT token from Authorization header
   *
   * @param header - Authorization header value (e.g., "Bearer eyJhbGc...")
   * @returns Token string if valid format, null otherwise
   *
   * Supports case-insensitive "Bearer" prefix and trims whitespace
   */
  extractTokenFromHeader(header: string): string | null {
    if (!header || header.trim() === '') {
      return null;
    }

    const trimmedHeader = header.trim();

    // Case-insensitive check for "Bearer " prefix
    if (!trimmedHeader.toLowerCase().startsWith(this.BEARER_PREFIX.toLowerCase())) {
      return null;
    }

    const token = trimmedHeader.slice(this.BEARER_PREFIX.length).trim();

    return token || null;
  }
}
